extensions/typescript-language-features/src/utils/previewer.ts

```diff
@@ -2,29 +2,119 @@
  *  Copyright (c) Microsoft Corporation. All rights reserved.
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
-
+import * as path from 'path';
 import * as vscode from 'vscode';
 import type * as Proto from '../protocol';
 
-function replaceLinks(text: string): string {
-	return text
-		// Http(s) links
-		.replace(/\{@(link|linkplain|linkcode) (https?:\/\/[^ |}]+?)(?:[| ]([^{}\n]+?))?\}/gi, (_, tag: string, link: string, text?: string) => {
+function regexConcat(re: RegExp[], flags?: string, sep: string = '') {
+	return new RegExp(re.map(re => re.source).join(sep), flags);
+}
+
+const LINK_SEARCHES = [
+	/\{@(link|linkplain|linkcode)\s+([^}]*)\}/, // {@link <link>}
+	/(\[[^\]]*\]\([^)]*\))/, // match markdown url so we dont replace it twice [<text>](<link>)
+	/(?=(?:file):\/\/)([^\s{}\)\[\]]+)/, // match supported top level links
+];
+
+const LINK_REGEXP = regexConcat(LINK_SEARCHES, 'gi', '|');
+
+function parseLinkMatch(match: unknown, definition?: Proto.DefinitionResponse['body']): [null, string?] | [string, string, string?] {
```

Comment:
We also do not want to add complex link parsing logic to the VS Code extension as this would not benefit any other consumers of the TS service. Instead the parsing should likely happen on the TypeScript side, with VS Code simply transforming the returned data into a usable format

Reply:
I am not sure this is really much more complex when all is said and done.  This is already parsing for `{@link}` so this is simply adding conditions to that same parsing which should not add much overhead unless it gets a match. 
 The condition of text being matched means that it is content that fits this format anyway. 

You should probably not support linking at all if it doesn't really work.  At least just ignore relative urls as well as {@link} because:

1. @link will never work when cmd+clicking them if they have a label since it will try to open the wrong url but in the hover it will be correct, which means the behavior is not only inconsistent but broken.
  - Another option is to not support the `|` syntax at all for link and just have the `{@link url this is the text}` which this PR does make work.
2. relative urls will only work in some case, but not most such as `../` or when rendered from another context.
3. file:// urls also do not work currently when relative and using @link 
4. file:// urls will never work when using markdown `[test](file://./plugins.ts)`

Probably should just remove relative link support completely so-as not to cause problems since at the moment it will fail almost 100% of the time.

Reply:
The existing `@link` parsing was added as a quick solution that addresses the simplest case. `{@link http://example.com}` is what the majority `@link` uses I've seen look like.

Without more user feedback, I'm also not convinced there is enough demand for file://./`style link support. Such links just look wrong to me and I also don't know if other tooling would support these